### DISCUSSION TITLE
멀티에서 클라이언트 캐릭터가 텔레포트 할 때 빠른속도로 걸어가는 것처럼 보입니다.

### DISCUSSION THREAD

**Discussion by jgw1413** [naverz/zepeto-studio-kor#1129] [5. Q&A]
현재 텔레포트 정보를 인터페이스로 만들어서 클라이언트 캐릭터가 텔레포트 했을 때 핸들러로 텔레포트 되게 구현했습니다.

클라이언트가 텔레포트 하면 클라이언트 캐릭터 State가 Teleport로 변경되는 것까지 확인 됐습니다.

근데 바로 텔레포트가 되는게 아니라 빠른속도로 뛰어가는것처럼 보입니다...

혹시 해결 방법이 있을까요? ㅠㅠ

**Comment by ZdevMonZ** [comment]
https://github.com/naverz/zepeto-studio-kor/discussions/1090 여기 답변 참고해서 저도 구현 성공했어요 ㅎ

**Reply by jgw1413** [reply]
저두 이거 참고해서 구현했지만, 서버전송해서 클라이언트 캐릭터가 Teleport 한것은 인식했지만 빠른속도로 뛰어가더라구요 ㅠㅠ...
맵 이동, 의자 앉기 제스쳐에 꼭 활용해야하는데 어디가 문제인지 감이 안잡히네요..
아래는 ClientStarterV2.ts의 RoomJoined안에 핸들러 구현한 코드입니다.

 // 텔레포트 메세지 핸들러 추가
            room.AddMessageHandler("teleport", (message: TeleportInfo) => {
                let teleportMessage = {sessionId:message.sessionId, teleportPoint:message.teleportPoint} as TeleportInfo;
                ZepetoPlayers.instance.GetPlayer(teleportMessage.sessionId).character.Teleport(teleportMessage.teleportPoint.position, teleportMessage.teleportPoint.rotation);
            });

**Reply by ZdevMonZ** [reply]
실제로 제페토 캐릭터를 텔레포트하는 코드는 넣으셨어요?

**Reply by jgw1413** [reply]
Teleport.ts라는 스크립트에서

1. OnTriggerEnter 에서 로컬플레이어 체크

2. 확인된 로컬플레이어는 정해둔 Transform으로 이동

/// Teleport.ts

    OnTriggerEnter(collider: UnityEngine.Collider)
    {
        if(collider.GetComponent<ZepetoCharacter>() === this.zepetoCharacter)
        {        
            if(this.zepetoCharacter.CurrentState !== CharacterState.Teleport)
            {
                this.zepetoCharacter.Teleport(this.teleportPoint.position, this.teleportPoint.rotation);

                if(this.clientStarterV2 !== null) {
                    this.clientStarterV2.SendTeleportState(this.teleportPoint);
                }
            }
        }
    }

3. 서버에 텔레포트 메세지 전달

/// ClientStarterV2.ts

    // 텔레포트 정보 Send
    public SendTeleportState(transform: UnityEngine.Transform)
    {
        // 이동 할 텔레포트 트랜스폼
        this.room.Send("teleport", transform);
    }

현재 이렇게 구현되어 있습니다! 다른 클라이언트 텔레포트 한것은 체크가 되는데 그 위치로 순간이동이 아닌 빠르게 이동을 하더라구요 ㅠㅠ

**Reply by jgw1413** [reply]
![캡처](https://user-images.githubusercontent.com/43944952/183278757-806dcf60-14da-45a8-986d-0fcd25bced92.PNG)

이 사진은 ClientStarterV2 Start() 함수 내의 텔레포트 메세지 구현 부분입니다

**Comment by koldshine** [comment]
두가지를 확인해봐야 할듯 합니다
1.서버에서 "teleport"값으로 broadcast 하는지
2.클라에서 다른 캐릭터가 teleport상태일때 이동 동기화를 중단하는지

**Reply by jgw1413** [reply]
덕분에 힌트 얻고 비슷하게 해결했습니다 감사합니다 ^^
